container/libcontainer/handler.go

```diff
@@ -758,6 +756,89 @@ func (h *Handler) GetProcesses() ([]int, error) {
 	return pids, nil
 }
 
+// ResetWss to be run as gorutine for non blocking referenced bytes reset
+func (h *Handler) ResetWss(containerName string) {
+
+	if *referencedResetInterval == time.Duration(0) {
+		return
+	}
+	castResetInterval := *referencedResetInterval
+	time.Sleep(time.Duration(rand.Intn(int(castResetInterval.Seconds()))))
+
+	for {
+		select {
+		case <-h.ReferencedResetStopper:
+			klog.V(5).Infof("Finished reseting wss for %s", containerName)
+			return
+		default:
+			pids, err := h.cgroupManager.GetPids()
+			if err != nil {
+				klog.V(5).Infof("Reset failed to collect pids for handler %s", containerName)
+
+			}
+			if len(pids) == 0 && err == nil {
+				klog.V(5).Infof("Tried to reset wss for container without PIDs %s", containerName)
+
+			}
+			start := time.Now()
+			h.referencedMemoryMutex.Lock()
+			err = clearReferencedBytes(pids)
+			if err != nil {
+				klog.Warningf("Failed clearing referenced bytes for %s", containerName)
+			}
+			h.referencedMemoryMutex.Unlock()
+			elapsed := time.Since(start)
+			if elapsed > *referencedResetInterval {
+				klog.Warningf("Exceeded referenced reset interval for %s", containerName)
+				elapsed = elapsed % *referencedResetInterval
+			}
+			toSleep := *referencedResetInterval - elapsed
+			time.Sleep(toSleep)
+		}
+	}
+}
+
+// ReadSmaps to be run as gorutine for non blocking referenced bytes read
+func (h *Handler) ReadSmaps(containerName string) {
+	klog.V(5).Infof("Starting WSS collection for %s", containerName)
+	if *referencedReadInterval == 0 {
+		return
+	}
+	castReadInterval := *referencedReadInterval
+	time.Sleep(time.Duration(rand.Intn(int(castReadInterval.Seconds()))))
```

Comment:
Could you explain this sleep?

Reply:
Both of those sleeps are responsible for preventing cpu usage spikes. Imagine a situation that cadvisor is started on a machine populated by a lot of containers, or a situation in which the containers are created 'in a batch' that is a lot of containers in a short period of time. In any of those cases, with every container creation a new handler starts, and so does the thread responsible for reading smaps file(and if requestested, another one for reseting it). 

As the interval is common for all the threads(it comes from parameter) now you are in situation, where all of those reading threads call their respective smaps files at almost the same moment. As smaps is not a regular file but an interface to a kernel function, this will trigger kernel execution and eat up cpu. Have a lot of not-so-multiprocess containers and you'll find yourself having cpu usage spikes every `referenced_read_interval`